File: src/com/suushiemaniac/cubing/bld/filter/condition/IntCondition.java
package com.suushiemaniac.cubing.bld.filter.condition;

public class IntCondition {
    public static IntCondition EXACT(int point) {
        return new IntCondition(point, point);
    }

    public static IntCondition PLUSMINUS(int point, int plusMinus) {
        return new IntCondition(point - plusMinus, point + plusMinus);
    }

    public static IntCondition INTERVAL(int min, int max) {
        return new IntCondition(min, max);
    }

    public static IntCondition MINIMUM(int min) {
        return new IntCondition(min, Integer.MAX_VALUE);
    }

    public static IntCondition MAXIMUM(int max) {
        return new IntCondition(Integer.MIN_VALUE, max);
    }

    public static IntCondition ANY() {
        return new IntCondition(Integer.MIN_VALUE, Integer.MAX_VALUE);
    }

    public static IntCondition NONE() {
        return new IntCondition(0, 0);
    }

    private int min, max;

    private IntCondition(int min, int max) {
        this.min = min;
        this.max = max;
    }

    public int getMin() {
        return this.min;
    }

    public int getMax() {
        return this.max;
    }

    public void capMin(int minCap) {
        if (this.min < minCap) this.setMin(minCap);
    }

    public void capMax(int maxCap) {
        if (this.max > maxCap) this.setMax(maxCap);
    }

    public boolean evaluate(int compareTo) {
        return this.min <= compareTo && compareTo <= this.max;
    }

    public boolean isPrecise() {
        return this.min == this.max;
    }

    public int getInterval() {
        return this.max - this.min;
    }

    public float getAverage() {
        return (this.min + this.max) / 2.f;
    }

    public void setMax(int max) {
        this.max = max;
        if (this.min > max) this.min = max;
    }

    public void setMin(int min) {
        this.min = min;
        if (this.max < min) this.max = min;
    }

    @Override
    public String toString() {
        return this.min + ":" + this.max;
    }
}

File: src/com/suushiemaniac/cubing/bld/filter/condition/IntCondition.kt
package com.suushiemaniac.cubing.bld.filter.condition

data class IntCondition(private var min: Int, private var max: Int) {
    val isPrecise: Boolean
        get() = this.min == this.max

    val interval: Int
        get() = this.max - this.min

    val average: Float
        get() = (this.min + this.max) / 2f

    fun getMin(): Int {
        return this.min
    }

    fun getMax(): Int {
        return this.max
    }

    fun capMin(minCap: Int) {
        if (this.min < minCap) this.setMin(minCap)
    }

    fun capMax(maxCap: Int) {
        if (this.max > maxCap) this.setMax(maxCap)
    }

    fun evaluate(compareTo: Int): Boolean {
        return this.min <= compareTo && compareTo <= this.max
    }

    fun setMax(max: Int) {
        this.max = max
        if (this.min > max) this.min = max
    }

    fun setMin(min: Int) {
        this.min = min
        if (this.max < min) this.max = min
    }

    override fun toString(): String {
        return this.toString(":")
    }

    fun toString(delim: String): String {
        return this.min.toString() + delim + this.max.toString()
    }

    companion object {
        fun EXACT(point: Int): IntCondition {
            return IntCondition(point, point)
        }

        fun POINT(point: Int): IntCondition {
            return IntCondition.EXACT(point)
        }

        fun PLUSMINUS(point: Int, plusMinus: Int): IntCondition {
            return IntCondition(point - plusMinus, point + plusMinus)
        }

        fun INTERVAL(min: Int, max: Int): IntCondition {
            return IntCondition(min, max)
        }

        fun BETWEEN(min: Int, max: Int): IntCondition {
            return IntCondition.INTERVAL(min, max)
        }

        fun MINIMUM(min: Int): IntCondition {
            return IntCondition(min, Integer.MAX_VALUE)
        }

        fun MIN(min: Int): IntCondition {
            return IntCondition.MINIMUM(min)
        }

        fun MAXIMUM(max: Int): IntCondition {
            return IntCondition(Integer.MIN_VALUE, max)
        }

        fun MAX(max: Int): IntCondition {
            return IntCondition.MAXIMUM(max)
        }

        fun ANY(): IntCondition {
            return IntCondition(Integer.MIN_VALUE, Integer.MAX_VALUE)
        }

        fun NONE(): IntCondition {
            return IntCondition.EXACT(0)
        }
    }
}
